File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/ClientApi/Session/LoadingEntities.java
import net.ravendb.client.documents.CloseableIterator;
import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.commands.StreamResult;
import net.ravendb.client.documents.session.*;
import net.ravendb.client.documents.session.loaders.ILoaderWithInclude;
import net.ravendb.client.primitives.Reference;

import java.io.ByteArrayOutputStream;
import java.util.Collection;
import java.util.Map;

public class LoadingEntities {
    private interface IFoo {
        //region loading_entities_1_0
        <T> T load(Class<T> clazz, String id);
        //endregion

        //region loading_entities_2_0
        ILoaderWithInclude include(String path);

        <TResult> Map<String, TResult> load(Class<TResult> clazz, String... ids);

        <TResult> Map<String, TResult> load(Class<TResult> clazz, Collection<String> ids);

        <TResult> TResult load(Class<TResult> clazz, String id);
        //endregion

        //region loading_entities_3_0
        <TResult> Map<String, TResult> load(Class<TResult> clazz, String... ids);

        <TResult> Map<String, TResult> load(Class<TResult> clazz, Collection<String> ids);
        //endregion

        //region loading_entities_4_0
        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix);

        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix, String matches);

        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix, String matches, int start);

        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix, String matches, int start, int pageSize);

        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix, String matches, int start, int pageSize, String exclude);

        <T> T[] loadStartingWith(Class<T> clazz, String idPrefix, String matches, int start, int pageSize, String exclude, String startAfter);
        //endregion

        //region loading_entities_5_0
        <T> CloseableIterator<StreamResult<T>> stream(IDocumentQuery<T> query);

        <T> CloseableIterator<StreamResult<T>> stream(IDocumentQuery<T> query, Reference<StreamQueryStatistics> streamQueryStats);

        <T> CloseableIterator<StreamResult<T>> stream(IRawDocumentQuery<T> query);

        <T> CloseableIterator<StreamResult<T>> stream(IRawDocumentQuery<T> query, Reference<StreamQueryStatistics> streamQueryStats);

        <T> CloseableIterator<StreamResult<T>> stream(Class<T> clazz, String startsWith);

        <T> CloseableIterator<StreamResult<T>> stream(Class<T> clazz, String startsWith, String matches);

        <T> CloseableIterator<StreamResult<T>> stream(Class<T> clazz, String startsWith, String matches, int start);

        <T> CloseableIterator<StreamResult<T>> stream(Class<T> clazz, String startsWith, String matches, int start, int pageSize);

        <T> CloseableIterator<StreamResult<T>> stream(Class<T> clazz, String startsWith, String matches, int start, int pageSize, String startAfter);
        //endregion

        //region loading_entities_6_0
        boolean isLoaded(String id);
        //endregion

        //region loading_entities_7_0
        <T> ConditionalLoadResult<T> conditionalLoad(Class<T> clazz, String id, String changeVector)
        //endregion
    }

    private static class Employee {

    }
    private static class Supplier {

    }
    private static class Product {
        private String supplier;

        public String getSupplier() {
            return supplier;
        }

        public void setSupplier(String supplier) {
            this.supplier = supplier;
        }
    }

    public LoadingEntities() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_1_1
                Employee employee = session.load(Employee.class, "employees/1");
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_2_1
                // loading 'products/1'
                // including document found in 'supplier' property
                Product product = session
                    .include("Supplier")
                    .load(Product.class, "products/1");

                Supplier supplier = session.load(Supplier.class, product.getSupplier()); // this will not make server call
                //endregion
                Product product1 = session.load()


            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_2_2
                // loading 'products/1'
                // including document found in 'supplier' property
                Product product = session
                    .include("Supplier")
                    .load(Product.class, "products/1");

                Supplier supplier = session.load(Supplier.class, product.getSupplier()); // this will not make server call
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_3_1
                Map<String, Employee> employees
                    = session.load(Employee.class,
                    "employees/1", "employees/2", "employees/3");
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_4_1
                // return up to 128 entities with Id that starts with 'employees'
                Employee[] result = session
                    .advanced()
                    .loadStartingWith(Employee.class, "employees/", null, 0, 128);
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_4_2
                // return up to 128 entities with Id that starts with 'employees/'
                // and rest of the key begins with "1" or "2" e.g. employees/10, employees/25
                Employee[] result = session
                    .advanced()
                    .loadStartingWith(Employee.class, "employees/", "1*|2*", 0, 128);
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_5_1
                try (CloseableIterator<StreamResult<Employee>> iterator =
                         session.advanced().stream(Employee.class, "employees/")) {
                    while (iterator.hasNext()) {
                        StreamResult<Employee> employee = iterator.next();
                    }
                }
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_5_2
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                session
                    .advanced()
                    .loadStartingWithIntoStream("employees/", baos);
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region loading_entities_6_1
                boolean isLoaded = session.advanced().isLoaded("employees/1");//false
                Employee employee = session.load(Employee.class, "employees/1");
                isLoaded = session.advanced().isLoaded("employees/1"); // true
                //endregion
            }

            User user;
            string changeVector;
            //region loading_entities_7_1
            try (IDocumentSession session = store.openSession()) {

                string changeVector;
                User user = new User("Bob");

                session.store(User, "users/1");
                session.saveChanges();

                changeVector = session.advanced().getChangeVectorFor(User user);
            }

            try (IDocumentSession session = store.openSession()) {
                // New session which does not track our User entity

                    // The given change vector matches
                    // the server-side change vector
                    // Does not load the document
                    ConditionalLoadResult<T> result1 = session.advanced()
                     .conditionalLoad(User.class,"users/1", changeVector);

                    // Modify the document
                    user.setName = "Bob Smith";
                    session.store(user);
                    session.saveChanges();

                    // Change vectors do not match
                    // Loads the document
                    ConditionalLoadResult<T> result2 = session.advanced().conditionalLoad(
                        User.class, "users/1", changeVector);
                    }

                //endregion
            }
        }
    }
}

File: Documentation/5.2/Samples/nodejs/ClientApi/Session/loadingEntities.js
import * as fs from "fs";
import { DocumentStore } from "ravendb";

let query, 
    id, 
    callback, 
    statsCallback, 
    documentType, 
    idsArray, 
    includes, 
    path,
    idPrefix,           
    matches,
    start,
    pageSize,
    exclude,
    startAfter,
    options;

const store = new DocumentStore();
const session = store.openSession();

//region loading_entities_1_0
await session.load(id, [documentType]);
//endregion

//region loading_entities_2_0
await session.include(path);
//endregion

//region loading_entities_3_0
await session.load(idsArray, [documentType]);
await session.load(idsArray, [options]);
//endregion

//region loading_entities_4_0
await session.advanced.loadStartingWith(idPrefix, [options]);

await session.advanced.loadStartingWithIntoStream(idPrefix, output, [options]);
//endregion

//region loading_entities_5_0
// stream query results
await session.stream(query, [statsCallback]);

// stream documents with ID starting with
await session.stream(idPrefix, [options]);
//endregion

//region loading_entities_6_0
await session.advanced.isLoaded(id);
//endregion

class Employee {

}
class Supplier {

}
class Product {
    constructor(supplier) {
        this.supplier = supplier;
    }
}

async function examples() {

    //region loading_entities_1_1
    const employee = await session.load("employees/1");
    //endregion

    //region loading_entities_2_1
    // loading 'products/1'
    // including document found in 'supplier' property
    const product = await session
        .include("supplier")
        .load("products/1");

    const supplier = await session.load(product.supplier); // this will *not* make a server call
    //endregion

    {
        //region loading_entities_2_2
        // loading 'products/1'
        // including document found in 'supplier' property
        const product = await session
            .include("supplier")
            .load("products/1");

        const supplier = await session.load(product.supplier); // this will *not* make a server call
        //endregion
    }

    //region loading_entities_3_1
    const employees = await session.load(
        [ "employees/1", "employees/2", "employees/3" ]);
    // {
    //     "employees/1": { ... },
    //     "employees/2": { ... }
    //     "employees/3": { ... }
    // }
    //endregion

    //region loading_entities_4_1
    // return up to 128 entities with Id that starts with 'employees'
    const result = await session
        .advanced
        .loadStartingWith("employees/", {
            start: 0, 
            pageSize: 128
        });
    //endregion

    {
        //region loading_entities_4_2
        // return up to 128 entities with Id that starts with 'employees/'
        // and rest of the key begins with "1" or "2" e.g. employees/10, employees/25
        const result = await session
            .advanced
            .loadStartingWith("employees/", {
                matches: "1*|2*",
                start: 0, 
                pageSize: 128
            });
        //endregion
    }

    //region loading_entities_5_1
    // stream() returns a Node.js Readable
    const stream = await session.advanced.stream("employees/");

    stream.on("data", data => {
        // Employee { name: 'Anna', id: 'employees/1-A' }
    });

    stream.on("error", err => {
        // handle errors
    });

    stream.on("end", () => {
        // stream ended
    });
    //endregion

    //region loading_entities_5_2
    const employeesFile = fs.createWriteStream("employees.json");
    await session.advanced.loadStartingWithIntoStream("employees/", employeesFile);
    //endregion

    {
        //region loading_entities_6_1
        session.advanced.isLoaded("employees/1"); // false
        const employee = await session.load("employees/1");
        session.advanced.isLoaded("employees/1"); // true
        //endregion
    }
}

{
    let id , object, changeVector;
    class User {}
    //region loading_entities_7_0
    await session.advanced.conditionalLoad(id, changeVector, object);
    //endregion

    //region loading_entities_7_1
    let session = store.openSession();
    await session.store(User, "users/1");
    await session.saveChanges();

    const _changeVector = session.advanced.getChangeVectorFor(User);

    // New session which does not track our User entity
    // The given change vector matches 
    // the server-side change vector
    // Does not load the document
    let session_2 = store.openSession();
    let user = await session.advanced
        .conditionalLoad("users/1", _changeVector,User);

    // Modify the document
    user.name = "Bob Smith";
    await session_2.store(User);
    await session_2.saveChanges();

    // Change vectors do not match
    // Loads the document
    let user2 = await session_2.advanced
        .conditionalLoad("users/1", _changeVector,User);
    //endregion
}

